File: KargoPaylasimJava/app/src/main/java/com/example/kargopaylasimjava/factory/VMFactories.java
package com.example.kargopaylasimjava.factory;


import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.example.kargopaylasimjava.repo.CargoRepository;
import com.example.kargopaylasimjava.service.TokenStore;
import com.example.kargopaylasimjava.viewmodel.AddressListViewModel;
import com.example.kargopaylasimjava.viewmodel.AddressViewModel;
import com.example.kargopaylasimjava.viewmodel.AuthViewModel;
import com.example.kargopaylasimjava.viewmodel.ShipmentViewModel;

public class VMFactories {

    public static class AuthVmFactory implements ViewModelProvider.Factory {
        private final CargoRepository repo;
        private final TokenStore tokenStore;

        public AuthVmFactory(CargoRepository repo, TokenStore tokenStore) {
            this.repo = repo;
            this.tokenStore = tokenStore;
        }

        @NonNull @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new AuthViewModel(repo, tokenStore);
        }
    }

    public static class ShipmentVmFactory implements ViewModelProvider.Factory {
        private final CargoRepository repo;
        public ShipmentVmFactory(CargoRepository repo) { this.repo = repo; }

        @NonNull @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new ShipmentViewModel(repo);
        }
    }

    public static class AddressVmFactory implements ViewModelProvider.Factory {
        private final CargoRepository repo;
        public AddressVmFactory(CargoRepository repo) { this.repo = repo; }

        @NonNull @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new AddressViewModel(repo);
        }
    }

    public static class AddressListVmFactory implements ViewModelProvider.Factory {
        private final CargoRepository repo;
        public AddressListVmFactory(CargoRepository repo) { this.repo = repo; }

        @NonNull @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            if (modelClass.isAssignableFrom(AddressListViewModel.class)) {
                return (T) new AddressListViewModel(repo);
            }
            throw new IllegalArgumentException("Unknown ViewModel class");
        }
    }
}

File: KargoPaylasimKotlin/app/src/main/java/com/example/kargopaylasimkotlin/factory/VMFactories.kt
package com.example.kargopaylasimkotlin.factory

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.kargopaylasimkotlin.repo.CargoRepository
import com.example.kargopaylasimkotlin.service.TokenStore
import com.example.kargopaylasimkotlin.viewmodel.AddressListViewModel
import com.example.kargopaylasimkotlin.viewmodel.AddressViewModel
import com.example.kargopaylasimkotlin.viewmodel.AuthViewModel
import com.example.kargopaylasimkotlin.viewmodel.ShipmentViewModel


class AuthVmFactory(private val repo: CargoRepository, private val tokenStore: TokenStore) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return AuthViewModel(repo, tokenStore) as T
    }
}

class ShipmentVmFactory(private val repo: CargoRepository) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return ShipmentViewModel(repo) as T
    }
}
class AddressVmFactory(private val repo: CargoRepository) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return AddressViewModel(repo) as T
    }
}
class AddressListVmFactory(private val repo: CargoRepository) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(AddressListViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return AddressListViewModel(repo) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}
